flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sink/AbstractWriterOperator.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.operators.sink;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.sink.Sink;
+import org.apache.flink.api.connector.sink.Writer;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.List;
+
+@Internal
+abstract class AbstractWriterOperator<InputT, CommT> extends AbstractStreamOperator<CommT>
+	implements OneInputStreamOperator<InputT, CommT>, BoundedOneInput {
+
+	private static final long serialVersionUID = 1L;
+
+	// ------------------------------- runtime fields ---------------------------------------
+
+	private transient Long currentWatermark;
+
+	private transient StreamRecord<CommT> streamRecord;
+
+	private transient Writer<InputT, CommT, ?> writer;
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		this.currentWatermark = Long.MIN_VALUE;
+
+		this.streamRecord = new StreamRecord<>(null);
+
+		writer = createWriter();
+	}
+
+	@Override
+	public void processElement(StreamRecord<InputT> element) throws Exception {
+		writer.write(element.getValue(), new Writer.Context() {
```

Comment:
Alternatively, we could keep a context object with mutable fields that we set. With this we would create less objects although I like the cleanliness that comes with creating a new context for each record.

I don't know what the runtime overhead of creating a new context per record is these days...

Reply:
NP！